store/search/src/main/java/org/apache/carbondata/store/master/Master.java

```diff
@@ -0,0 +1,279 @@
+    // add these blocks to the SearchRequest and fire the RPC call
+    Map<String, List<Distributable>> nodeBlockMapping = pruneBlock(table, columns, filter);
+
+    List<ListenableFuture<SearchResult>> futures = new ArrayList<>(nodeBlockMapping.size());
+
+    for (Map.Entry<String, List<Distributable>> entry : nodeBlockMapping.entrySet()) {
+      String hostname = entry.getKey();
+      List<Distributable> blocks = entry.getValue();
+      CarbonMultiBlockSplit mbSplit = new CarbonMultiBlockSplit(blocks, hostname);
+      ByteArrayOutputStream stream = new ByteArrayOutputStream();
+      DataOutput dataOutput = new DataOutputStream(stream);
+      mbSplit.write(dataOutput);
+      builder.setSplits(ByteString.copyFrom(stream.toByteArray()));
+
+      SearchRequest request = builder.build();
+
+      // do RPC to worker asynchronously and concurrently
+      ListenableFuture<SearchResult> future = workers.get(hostname).search(request);
+      futures.add(future);
+    }
+
+    // get all results from RPC response and return to caller
+    List<CarbonRow> output = new LinkedList<>();
+    for (ListenableFuture<SearchResult> future : futures) {
+      SearchResult result = future.get();
+      if (result.getQueryId() != queryId) {
+        throw new IOException(String.format(
+            "queryId in response does not match request: %d != %d", result.getQueryId(), queryId));
+      }
+      collectResult(result, output);
+    }
+    return output.toArray(new CarbonRow[output.size()]);
+  }
+
+  /**
+   * Prune data by using CarbonInputFormat.getSplit
+   * Return a mapping of hostname to list of block
+   */
+  private Map<String, List<Distributable>> pruneBlock(CarbonTable table, String[] columns,
+      Expression filter) throws IOException, InvalidConfigurationException {
+    JobConf jobConf = new JobConf(new Configuration());
+    Job job = new Job(jobConf);
+    CarbonTableInputFormat<Object> format = CarbonInputFormatUtil.createCarbonTableInputFormat(
+        job, table, columns, filter, null, null);
+
+    List<InputSplit> splits = format.getSplits(job);
+    List<Distributable> distributables = new ArrayList<>(splits.size());
+    for (InputSplit split : splits) {
+      distributables.add(((CarbonInputSplit)split));
+    }
+    return CarbonLoaderUtil.nodeBlockMapping(
+        distributables, -1, new ArrayList<String>(workers.keySet()),
+        CarbonLoaderUtil.BlockAssignmentStrategy.BLOCK_NUM_FIRST);
+  }
+
+  /**
+   * Fill result row to {@param output}
+   */
+  private void collectResult(SearchResult result,  List<CarbonRow> output) throws IOException {
+    for (ByteString bytes : result.getRowList()) {
+      CarbonRow row = GrpcSerdes.deserialize(bytes);
+      output.add(row);
+    }
+  }
+
+  /** return hostname of all workers */
+  public Set<String> getWorkers() {
+    return workers.keySet();
+  }
+
+  public static void main(String[] args) throws IOException, InterruptedException {
```

Comment:
what is the use of main here? is it for testing ?

Reply:
yes